### DISCUSSION TITLE
with-out-parameters crashing with posix_spawn_file_actions_init?

### DISCUSSION THREAD

**Discussion by catb0t** [factor/factor#2502] [Q&A]
I'm not sure if this is my error or a bug, so I'm not making it an issue yet.

If you run the following, Factor will say "Memory protection fault at address 0x0", and then crash on the next gc: 

```factor
USING: alien.syntax alien.data ;
: <posix-spawn-file-actions ( -- error actions ) 
    { pointer: posix_spawn_file_actions_t } 
    [ posix_spawn_file_actions_init ] with-out-parameters ;
```

```
fatal_error: Memory protection fault during gc: 0x164ae

Nursery   : [0x7f02e6180000 -> 0x7f02e6380000]    2097152 bytes
Aging     : [0x7f02e5d80000 -> 0x7f02e6180000]    4194304 bytes
Tenured   : [0x7f02d9980000 -> 0x7f02e5980000]  201326592 bytes
Cards     : [0x7f02d98ab010 -> 0x7f02d9975410]     828416 bytes

Contexts:

  Datastack   : [0x7f02e653f000 -> 0x7f02e657f000]     262144 bytes
  Retainstack : [0x7f02e64fd000 -> 0x7f02e653d000]     262144 bytes
  Callstack   : [0x7f02e63fb000 -> 0x7f02e64fb000]    1048576 bytes
```

The constructor for `posix_spawnattr_t` also breaks: 
```factor
USING: alien.syntax alien.data ;
: <posix-spawnattr> ( -- error attrp )
    { pointer: posix_spawnattr_t } [
      posix_spawnattr_init
    ] with-out-parameters ;
```
gives

```
Call stack underflow
```

Regardless of whether you use `pointer:` or not for the `c-type` in the array to `with-out-parameters`.

I am able to use these functions fine from C.

--- 

I have unfortunately been away from Factor since March as I got too busy working ): but wanted to move forward on the posix-spawn implementation. My knowledge of the FFI has diminished without exercise but I will learn again :) 

I forgot, I already had a branch going for posix-spawn; it has these constructors, but they have a trailing `throw-errno` clause which complicates matters, so I've omitted them here. [edit-posix_spawn](https://github.com/catb0t/factor/tree/edit-posix_spawn)

**Comment by mrjbq7** [comment]
Maybe you could manually use <ref> and deref

```factor
int <ref> [ posix_spawnattr_init ] keep int deref
```

or in this case maybe allocate a struct and then keep it

```factor
posix_spawnattr_t <struct> [ posix_spawnattr_init ] keep
```

or if that doesn't work, make a void pointer and keep it around as a pointer to the thing

```factor
void <ref> [ posix_spawnattr_init ] keep
```

**Reply by mrjbq7** [reply]
P.S., I'm totally fine with opening issues to track stuff like this.

**Reply by catb0t** [reply]
I could make this an issue since that seems like a problem with `with-out-parameters`. 

Your second suggestion works well to give a struct, I will use that for now.

**Comment by catb0t** [comment]
I see now this is likely just an issue with my struct layouts being wrong. Attempting to enumerate the members in Factor is a mistake because the actual contents are very unpredictable.

Better to leave them opaque `void*`s as they were.
